### PR TITLE
[Enhancement] support empty node prune

### PR DESCRIPTION
**Pull request by Seaven**
Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 24 / 24 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalScanOperator.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalValuesOperator.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSet.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSetType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/CastToEmptyRule.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/LimitPruneTabletsRule.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/RuleSet.java [hunk 1]
```diff
@@ -385,6 +385,21 @@ public class RuleSet {
                 OnlyJoinRule.getInstance()
         ));
 
+        REWRITE_RULES.put(RuleSetType.PRUNE_EMPTY_OPERATOR, ImmutableList.of(
+                PruneEmptyScanRule.OLAP_SCAN,
+                PruneEmptyScanRule.DELTALAKE_SCAN,
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/RuleSet.java:390]: currently deltalake table doesn't have partition prune

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/logical/LogicalIcebergScanOperator.java [hunk 2]
```diff
@@ -59,7 +59,7 @@ public void setScanOperatorPredicates(ScanOperatorPredicates predicates) {
     }
 
     @Override
-    protected boolean noPartitionSelected() {
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/logical/LogicalIcebergScanOperator.java:62]: This rule needs to be used after computeIcebergStatistics on iceberg table

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 152 / 163 (93.25%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalScanOperator.java|0|1|00.00%|[117]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneEmptyExceptRule.java|24|33|72.73%|[87, 88, 89, 90, 91, 93, 95, 97, 98]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneEmptyJoinRule.java|36|37|97.30%|[99]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalValuesOperator.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/RewriteTreeTask.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSet.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSetType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalOlapScanOperator.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneEmptyScanRule.java|14|14|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/ExistentialApply2OuterJoinRule.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/CastToEmptyRule.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneEmptyDirectRule.java|14|14|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneEmptyUnionRule.java|33|33|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/FeConstants.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneEmptyIntersectRule.java|8|8|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 152 / 163 (93.25%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalScanOperator.java|0|1|00.00%|[117]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneEmptyExceptRule.java|24|33|72.73%|[87, 88, 89, 90, 91, 93, 95, 97, 98]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneEmptyJoinRule.java|36|37|97.30%|[99]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalValuesOperator.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/RewriteTreeTask.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSet.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSetType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalOlapScanOperator.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneEmptyScanRule.java|14|14|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/ExistentialApply2OuterJoinRule.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/CastToEmptyRule.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneEmptyDirectRule.java|14|14|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneEmptyUnionRule.java|33|33|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/FeConstants.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneEmptyIntersectRule.java|8|8|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 152 / 162 (93.83%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneEmptyExceptRule.java|24|33|72.73%|[87, 88, 89, 90, 91, 93, 95, 97, 98]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneEmptyJoinRule.java|36|37|97.30%|[99]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalValuesOperator.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/RewriteTreeTask.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSet.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSetType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalOlapScanOperator.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneEmptyScanRule.java|14|14|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/ExistentialApply2OuterJoinRule.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/CastToEmptyRule.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneEmptyDirectRule.java|14|14|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneEmptyUnionRule.java|33|33|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/FeConstants.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneEmptyIntersectRule.java|8|8|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 152 / 162 (93.83%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneEmptyExceptRule.java|24|33|72.73%|[87, 88, 89, 90, 91, 93, 95, 97, 98]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneEmptyJoinRule.java|36|37|97.30%|[99]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalValuesOperator.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/RewriteTreeTask.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSet.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSetType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalOlapScanOperator.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneEmptyScanRule.java|14|14|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/ExistentialApply2OuterJoinRule.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/CastToEmptyRule.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneEmptyDirectRule.java|14|14|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneEmptyUnionRule.java|33|33|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/FeConstants.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneEmptyIntersectRule.java|8|8|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 152 / 162 (93.83%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneEmptyExceptRule.java|24|33|72.73%|[87, 88, 89, 90, 91, 93, 95, 97, 98]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneEmptyJoinRule.java|36|37|97.30%|[99]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalValuesOperator.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/RewriteTreeTask.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSet.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSetType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalOlapScanOperator.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneEmptyScanRule.java|14|14|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/ExistentialApply2OuterJoinRule.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/CastToEmptyRule.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneEmptyDirectRule.java|14|14|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneEmptyUnionRule.java|33|33|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/FeConstants.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneEmptyIntersectRule.java|8|8|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 157 / 167 (94.01%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneEmptyExceptRule.java|24|33|72.73%|[87, 88, 89, 90, 91, 93, 95, 97, 98]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneEmptyJoinRule.java|36|37|97.30%|[99]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalValuesOperator.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/RewriteTreeTask.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSet.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSetType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalOlapScanOperator.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneEmptyScanRule.java|14|14|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/ExistentialApply2OuterJoinRule.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/CastToEmptyRule.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneEmptyDirectRule.java|14|14|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneEmptyUnionRule.java|33|33|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/MvPlanContextBuilder.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/FeConstants.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneEmptyIntersectRule.java|8|8|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 157 / 167 (94.01%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneEmptyExceptRule.java|24|33|72.73%|[87, 88, 89, 90, 91, 93, 95, 97, 98]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneEmptyJoinRule.java|36|37|97.30%|[99]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalValuesOperator.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/RewriteTreeTask.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSet.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSetType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalOlapScanOperator.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneEmptyScanRule.java|14|14|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/ExistentialApply2OuterJoinRule.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/CastToEmptyRule.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneEmptyDirectRule.java|14|14|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneEmptyUnionRule.java|33|33|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/MvPlanContextBuilder.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/FeConstants.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneEmptyIntersectRule.java|8|8|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Review by kangkaisen — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/ExistentialApply2OuterJoinRule.java [hunk 3]
```diff
@@ -289,9 +289,7 @@ public List<OptExpression> transform2OuterJoin(OptimizerContext context, OptExpr
 
         // filter(UnCorrelation) agg -> project -> un-correlation filter
         if (null != apply.getPredicate()) {
-            OptExpression filterOptExpression =
-                    OptExpression.create(new LogicalFilterOperator(apply.getPredicate()), input.getInputs().get(1));
```
Comment (liuyehcf) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/ExistentialApply2OuterJoinRule.java:293]: so when to add `input.getInputs().get(1)`?

Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/ExistentialApply2OuterJoinRule.java:293]: duplicate code in line 299

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/RewriteTreeTask.java [hunk 4]
```diff
@@ -101,6 +100,10 @@ private boolean match(Pattern pattern, OptExpression root) {
             return false;
         }
 
+        if (pattern.children().size() > 0 && pattern.children().size() != root.getInputs().size() &&
+                pattern.children().stream().noneMatch(Pattern::isPatternMultiLeaf)) {
+            return false;
```
Comment (packy92) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/RewriteTreeTask.java:105]: Is this just for short circuit?

Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/RewriteTreeTask.java:105]: for strict check. pattern leaf could match multiple-children before this pr

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 155 / 165 (93.94%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneEmptyExceptRule.java|24|33|72.73%|[87, 88, 89, 90, 91, 93, 95, 97, 98]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneEmptyJoinRule.java|36|37|97.30%|[99]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalValuesOperator.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/RewriteTreeTask.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSet.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSetType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalOlapScanOperator.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneEmptyScanRule.java|14|14|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/ExistentialApply2OuterJoinRule.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/CastToEmptyRule.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneEmptyDirectRule.java|14|14|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneEmptyUnionRule.java|31|31|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/MvPlanContextBuilder.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/FeConstants.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneEmptyIntersectRule.java|8|8|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 31 / 32 (96.88%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalHudiScanOperator.java|0|1|00.00%|[65]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalScanOperator.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalValuesOperator.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/RewriteTreeTask.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSet.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSetType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalOlapScanOperator.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/CastToEmptyRule.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/LimitPruneTabletsRule.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Review by liuyehcf — APPROVED** [review]

**Review by packy92 — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.2

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.1

**Pull request merged**